### PR TITLE
Arrow Fx and Benchmarks

### PR DESCRIPTION
**Pull request by fedefernandez**
Created against the featured branch `arrow-fx-benchmark-2` that was cut off from the @raulraja's branch `rr-Fx-datatype`.

Migrates the [cats-effect AttemptBenchmark](https://github.com/typelevel/cats-effect/blob/master/benchmarks/shared/src/main/scala/cats/effect/benchmarks/AttemptBenchmark.scala) to Arrow. 

I guess the code can be improved, I'm not very familiar with the syntax.

-------------------------------------------
Process: 
- [x] Fix bug ForkFiber ~- and investigate why no law failed~
- [x] Green laws including Concurrent
- [x] Improve performance (make on par with arrow.effects.IO for a V1)
  - [x] DeepBind
  - [x] Pure

After:
  - [ ] Analyze `startCoroutine` alternatives
  - [ ] Improve performance further
  - [ ] Fx<E, A>
  - [ ] Fx<R, E, A>

     This was moved out of scope of this PR - WIP on fx-eval-on
  - [x] Rewrite `ContinueOn`/`UpdateContext` to `ContextSwitch` similar to `ConnectionSwitch`.
  - [ ] Add EventLoopCC to return back to original Thread.

### CODE REVIEW

**Comment by pakoito** [comment]
```
	ThrowsCount - [bracketCase] at /home/travis/build/arrow-kt/arrow/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/suspended/SuspendedSyntax.kt:94:3
	FunctionOnlyReturningConstant - [helloWorld] at /home/travis/build/arrow-kt/arrow/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/suspended/SuspendedSyntax.kt:221:1
	UnusedImports - [FxTest.kt] at /home/travis/build/arrow-kt/arrow/modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/FxTest.kt:4:1
	UnusedImports - [FxTest.kt] at /home/travis/build/arrow-kt/arrow/modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/FxTest.kt:7:1
```

Detekt lol Fix the imports!

**Comment by fedefernandez** [comment]
@pakoito @raulraja is it ok to merge? We're using a feature branch as base

**Comment by pakoito** [comment]
@raulraja ??

**Comment by raulraja** [comment]
it's fine to merge in the Fx branch but not on master. Simon and I are also finishing up changes to Fx

**Comment by raulraja** [comment]
talked to Fede and we are keeping this branch around to all work on it as they add the benchmarks so we can be testing them with the fx changes.

File: modules/benchmarks/arrow-benchmarks-effects/src/jmh/kotlin/arrow/benchmarks/Pure.kt [hunk 1]
```diff
@@ -24,38 +19,18 @@ open class Pure {
...
-  }
-
-  @Benchmark
-  fun envFx(): Int =
-    unsafe {
-      fxRunBlocking {
-        Fx { envFxPureLoop<Int, Int>(0)(0) }
-      }.getOrHandle { 0 }
-    }
+  private tailrec suspend fun fxPureLopp(i: Int): suspend () -> Int =
```
Comment (pakoito) [inline modules/benchmarks/arrow-benchmarks-effects/src/jmh/kotlin/arrow/benchmarks/Pure.kt]: loop*

File: modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/suspended/fx/FxFrame.kt [hunk 2]
```diff
@@ -37,8 +37,8 @@ internal interface FxFrame<in A, out B> : (A) -> B {
     fun <A> attempt(): (A) -> Fx<Either<Throwable, A>> = AttemptFx as (A) -> Fx<Either<Throwable, A>>
 
     private object AttemptFx : FxFrame<Any?, Fx<Either<Throwable, Any?>>> {
-      override fun invoke(a: Any?): Fx<Either<Throwable, Any?>> = Fx.Pure(Either.Right(a), 0)
-      override fun recover(e: Throwable): Fx<Either<Throwable, Any?>> = Fx.Pure(Either.Left(e), 0)
+      override inline fun invoke(a: Any?): Fx<Either<Throwable, Any?>> = Fx.Pure(Either.Right(a), 0)
```
Comment (nomisRev) [inline modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/suspended/fx/FxFrame.kt]: Is this taking into account when the interface methods are not inline?

**Comment by nomisRev** [comment]
This has all been implemented in separate PRs and has all been merged.

**Pull request closed without merging**